DESIGN.md

```diff
@@ -0,0 +1,79 @@
+- **DeleteJob**: receives a PID and returns a DeleteJobOutput
+- **ListJobs**: returns a JobList
+- **GetOutput**: receives a PID and returns a JobOutput
+
+### Datatypes
+- **Job**
+    - Command (string):
+    Command holds a valid shell command that is run on the underlying server shell.
+    - PID (integer):
+    PID holds an integer representing job ID. In this project it will correlate directly to the pid of the process running on the RPC server for a given job. In a production RPC system it is advisable to use a randomized PID.
+    - CreatedTime (string):
+    Timestamp that represents when the process was launched on the server
+- **JobList**
+    - Jobs ([]Job):
+    A list of Job obejcts as defined above in this list
+- **JobOutput**
+    - Command (string):
+    Holds the shell command as originally defined in the corresponding call to CreateJob
+    - CurrentOutput (string):
+    Current cummulative contents of stdout from Job
+    - Complete (boolean):
+    Boolean that represents whether the job is still running or if it has finished
+- **DeleteJobOutput**
+    - Job (Job):
+    Job object as defined above in this list
+    - Exit (integer):
+    Exit code as returned from the underlying process
+    - Interrupted (boolean):
+    Boolean is set to true if underlying process was still running when DeleteJob was called
+- **CreateJobOutput**
+    - Job (Job):
+    Created Job object as defined above in this list
+    - Success (boolean):
+    Boolean set to true if the job was launched successfully
+    - Error (string):
+    If success is set to False, an error can be found in Error
+- **PID**
+    Single integer representing the PID of the underlying process
+
+### Data flow
+Each handler will receive the endpoint input as well as a context object from which the client certificate and trust chain can be extracted. After the Authorization routine is applied to the CA certificate found in the client certificate trust chain, the handler will process the input and return the output.
+
+![flowchart for CreateJob API call](doc/data_flow.png)
+
+## Controller
+The backend will leverage the CommandContext type from the os/exec package. This will provide convenient facilities for cancelling running jobs and garnering job output. Additionally, it will handle passing through environment variables for us.
+
+A single datatype, JobCtl, will be defined to contain the following peices of data:
+- The os/exec CommandContext instance
+- The cancel function derived from teh context passed into the CommandContext instance
+- The "Job" instance as defined in the API datatypes.
+    
+A sync.Map will be used to cache JobCtl objects. The JobCtl objects will be mapped to their PID. The sync.Map will take care of synchronization, reducing the likelihood of concerrency errors like race conditions. Methods will be defined to retrieve API Job objects, create new JobCtls, delete existing JobCtls, and fetch output from a JobCtl.
```

Comment:
what's your opinion on using sync map and having to check type safety every time you read from it vs using a normal map with a sync mutex

Reply:
can you also comment on the `reducing the likelihood of concurrency errors like race conditions` part? Will it still be a race accessing the sync map or no?

Reply:
performance aside, you'd still want to wrap the sync map and not access it directly, to prevent storing keys of different types, type checking the values... because you wouldn't want do rewrite these checks every time you use the map in a different part of the code.

Since you're already writing a wrapper for the map, might as well include the sync mechanism yourself and do without all the type safety checks that are already included in normal maps.

you also omit this part of the documentation a couple of line above points you mentioned:
```
The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content. 
```

I am not against sync maps and not saying you shouldn't use them, it's just they're sometimes more work and have worse performance so I just wanted to get your opinion on that.

Reply:
> Since you're already writing a wrapper for the map, might as well include the sync mechanism yourself and do without all the type safety checks that are already included in normal maps.

This is a convincing argument. When thinking about the implementation of the wrapper I can imagine it would be easier to just work with a normal mutex instead of manage all of the type checking code. I'll rework this section of the design document accordingly.